sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -1178,4 +1178,36 @@ class SessionCatalog(
     }
   }
 
+  /**
+   * Get an identical copy of the `SessionCatalog`.
+   * The temporary tables and function registry are retained.
+   * The table relation cache will not be populated.
+   * @note `externalCatalog` and `globalTempViewManager` are from shared state, don't need deep copy
+   * `FunctionResourceLoader` is effectively stateless, also does not need deep copy.
+   * All arguments passed in should be associated with a particular `SparkSession`.
+   */
+  def copy(
+      conf: CatalystConf,
+      hadoopConf: Configuration,
+      functionRegistry: FunctionRegistry,
+      parser: ParserInterface): SessionCatalog = {
+
+    val catalog = new SessionCatalog(
+      externalCatalog,
+      globalTempViewManager,
+      functionResourceLoader,
+      functionRegistry,
+      conf,
+      hadoopConf,
+      parser)
+
+    synchronized {
```

Comment:
Right, I should mention we decided to keep it `synchronized`. `catalog` is not yet accessible outside local scope.